### ISSUE TITLE
luaoftload v3.17 with HarfBuzz does awful line breaking with Garamond

### ISSUE THREAD

**Issue by drunken-whaler** [latex3/luaotfload#175]
Reposting from [here](https://tex.stackexchange.com/q/581877/171234) by request.

I have this long document which I've been maintaining for years now, which I use Adobe's Garamond Premier Pro font for typesetting, and the whole document is fine-tuned to avoid overfull boxes and such (so I can't easily change fonts). Since the last `luaotfload` update, the whole document got completely messed up regarding line breaking. Took me some time to narrow down the issue to this example:
```latex
\documentclass[11pt]{article}
\usepackage{fontspec}
\setmainfont{Garamond Premier Pro}[Renderer=HarfBuzz]
\usepackage{lipsum}
\begin{document}
\lipsum[17]
\end{document}
```

And the problem requires that specific combination to show up. If I change `11pt` to `10pt`, line breaking looks right. If I remove `Renderer=HarfBuzz`, line breaking looks right.  If I change the font to, say, `EB Garamond`, line breaking looks right (or any font, in fact. I could only reproduce the problem with Garamond Premier Pro). Depending on the language I use (with `babel`), line breaking looks slightly better, but still off.

If I downgrade `luaotfload` to v3.15 (with `tlmgr restore luaotfload 56269`), line breaking looks great:

[![enter image description here][1]][1]

But if I update it again to v3.17 (with `tlmgr update luaotfload`), line breaking is a wreck:

[![enter image description here][2]][2]

What's going on? Is Garamond Premier Pro no longer supported by `luaotfload`?

P.S.: Sorry for requiring a commercial font in this question, but I really could not reproduce the problem with another font (I tried a few, but all worked correctly).

  [1]: https://i.stack.imgur.com/mQ88E.png
  [2]: https://i.stack.imgur.com/GezhK.png

**Comment by zauguin** [comment]
Thank you for the report.

> I have this long document which I've been maintaining for years now, which I use Adobe's Garamond Premier Pro font for typesetting, and the whole document is fine-tuned to avoid overfull boxes and such (so I can't easily change fonts).

I think you already heard that on StackExchange, but I would recommend the node shaper for this kind of use currently. For latin fonts the node shaper works great and it is more mature. There are some issues with the harf hyphenation support which are actively worked on, which means that hyphenation is rather likely to change between versions.

> What's going on? Is Garamond Premier Pro no longer supported by `luaotfload`?
> 
> P.S.: Sorry for requiring a commercial font in this question, but I really could not reproduce the problem with another font (I tried a few, but all worked correctly).

It's odd, we changed some hyphenation code but that shouldn't make a difference here. Sadly I can't really debug without the font. Could you send us the log file when running

```
\documentclass[11pt]{article}
\usepackage{fontspec}
\setmainfont{Garamond Premier Pro}[Renderer=HarfBuzz]
\usepackage{lipsum}
\begin{document}
\showoutput
\tracingparagraphs1
\lipsum[17]
\end{document}
```

**Comment by drunken-whaler** [comment]
Hi @zauguin, thank you for looking into this.

> There are some issues with the harf hyphenation support which are actively worked on, which means that hyphenation is rather likely to change between versions.

Sheesh, I guess that's the final nail on this coffin.  I really don't want my document to change, so I guess I'll have to switch to the default renderer.

> Could you send us the log file when running

Sure, here's the [log with v3.15](https://github.com/latex3/luaotfload/files/5927603/garbug_3.15.log) and the [with v3.17](https://github.com/latex3/luaotfload/files/5927626/garbug_3.17.log).

> Sadly I can't really debug without the font.

I can send you the `.otf` file that gets loaded in that example (it seems to be only one), just drop me a line at: drunken \<dot> whaler3 (gmail address).

**Comment by zauguin** [comment]
This seems to be an engine bug. It should be fixed if [my LuaTeX patch](https://mailman.ntg.nl/pipermail/dev-luatex/2021-February/006448.html) gets merged.

**Comment by drunken-whaler** [comment]
@zauguin Oh, looks like I kept you busy, huh?  Honestly, I didn't understand exactly what's going on in the patch you linked, but you managed to reproduce with stock LuaTeX and CM fonts, so it seems it's not only HarfBuzz that fails, right?  HarfBuzz only makes it _more likely_ to fail, am I right?

Anyway, thanks a million for the effort you put into this! Much appreciated

**Comment by zauguin** [comment]
@drunken-whaler Don't worry, I didn't expect you to understand the details of that, I only linked it to have a reference when trying to determine if this is fixed later. 

While it's not only HarfBuzz, fonts which use luaotfload's default renderer ("node") are not affected since they don't use the constructs which cause the issues. It does happen with traditional TeX fonts in LuaTeX though, but there the elements are often smaller and therefore it is less obvious. Basically we decided to use a feature in the HarfBuzz support code which wasn't used much before, so we shouldn't be that surprised that there were some issues.
